iXBRLViewerPlugin/xhtmlserialize.py

```diff
@@ -13,29 +13,172 @@
+            prefix = None
+        elif node.nsmap.get(node.prefix, None) == qname.namespace:
+            prefix = node.prefix
+        else:
+            prefix = next(iter(sorted((p for p, ns in nsmap.items() if ns == qname.namespace), key = self.prefix_sort)))
+        if prefix is None:
+            return qname.localname
+        return "%s:%s" % (prefix, qname.localname)
+
+    def qname_for_attr(self, tag, nsmap):
+        qname = etree.QName(tag)
+        if qname.namespace is None:
+            return qname.localname
+        if qname.namespace == XML_NS:
+            prefix = 'xml'
+        else:
+            prefix = next(iter(sorted((p for p, ns in nsmap.items() if ns == qname.namespace and p is not None), key = self.prefix_sort)))
+        return "%s:%s" % (prefix, qname.localname)
+
+    def is_selfclosable(self, n):
+        qname = etree.QName(n)
+        return (
+                qname.localname in XHTMLSerializer.SELF_CLOSABLE
+                and (
+                    qname.namespace == XHTML_NS
+                    or (self.assume_xhtml and qname.namespace is None)))
+
+    def escape_attr(self, s):
+        return re.sub(r'([<>"&\u0001-\u001F\u007F-\u009F])', lambda m: self.escape_char(m.group(0)), s)
+
+    def xmlns_declaration(self, prefix, uri):
+        if prefix is None:
+            return 'xmlns="%s"' % self.escape_attr(uri)
+        return 'xmlns:%s="%s"' % (prefix, self.escape_attr(uri))
+
+    def namespace_declarations(self, new_nsmap, cur_nsmap):
+        changed = set(p[0] for p in (set(new_nsmap.items()) ^ set(cur_nsmap.items())))
+        return sorted(self.xmlns_declaration(p, new_nsmap[p]) for p in changed)
+
+    def escape_char(self, c):
+        return self.ENTITIES.get(c, '&#x%2X;' % ord(c))
+
+    def write_escape_text(self, s, escape_mode):
+        if s is None:
+            return
+        if escape_mode == EscapeMode.STYLE:
+            # HTML does not understand XML escapes inside a style element.
+            # Don't escape ">".  Escaping isn't required by XML, but is common
+            # practice, so we do it elsewhere.
+            # "&" and "<" may only appear in a string or comment in CSS, so escape
+            # as if they're in a string.
+            self.write(re.sub(r'([<&])', lambda m: "\\%06X" % ord(m.group(1)), s))
+        else:
+            self.write(re.sub(r'([<>&\u0001-\u001F\u007F-\u009F])', lambda m: self.escape_char(m.group(0)),s))
+
+    def write_attributes(self, node):
+        for qname, value in sorted((self.qname_for_attr(k, node.nsmap), v) for k, v in node.items()):
+            self.write(' %s="' % qname)
+            self.write(self.escape_attr(value))
+            self.write("\"")
+
+    def write_comment(self, n, escape_mode):
+        self.write('<!--' + n.text + '-->')
+        self.write_escape_text(n.tail, escape_mode)
+
+    def write_processing_instruction(self, n, escape_mode):
+        self.write( '<?' + n.target + ' ' + n.text + '?>')
```

Comment:
It's odd that `<?pi?>` round-trips as `<?pi ?>`.  It might be possible to distinguish the two(?), but the former seems preferable.  I somehow doubt PIs are in regular use anyway :-).

Reply:
I don't think we can distinguish between `<?pi ?>` and `<?pi?>`, but we can with `<?pi foo="bar" ?>` and `<?pi foo="bar"?>`.

I've made it collapse `<?pi ?>` to `<?pi?>` and preserve space otherwise.